docker/docker_image_src.go

```diff
@@ -53,20 +54,42 @@ func newImageSource(ctx context.Context, sys *types.SystemContext, ref dockerRef
 	// non-mirror original location last; this both transparently handles the case
 	// of no mirrors configured, and ensures we return the error encountered when
 	// acessing the upstream location if all endpoints fail.
-	manifestLoadErr := errors.New("Internal error: newImageSource returned without trying any endpoint")
 	pullSources, err := registry.PullSourcesFromReference(ref.ref)
 	if err != nil {
 		return nil, err
 	}
+	type attempt struct {
+		ref reference.Named
+		err error
+	}
+	attempts := []attempt{}
 	for _, pullSource := range pullSources {
 		logrus.Debugf("Trying to pull %q", pullSource.Reference)
 		s, err := newImageSourceAttempt(ctx, sys, pullSource, primaryDomain)
 		if err == nil {
 			return s, nil
 		}
-		manifestLoadErr = err
+		attempts = append(attempts, attempt{
+			ref: pullSource.Reference,
+			err: err,
+		})
+	}
+	switch len(attempts) {
+	case 0:
+		return nil, errors.New("Internal error: newImageSource returned without trying any endpoint")
+	case 1:
+		return nil, attempts[0].err // If no mirrors are used, perfectly preserve the error type and add no noise.
+	default:
+		// Don’t just build a string, try to preserve the typed error.
+		primary := &attempts[len(attempts)-1]
+		extras := []string{}
+		for i := 0; i < len(attempts)-1; i++ {
+			// This is difficult to fit into a single-line string, when the error can contain arbitrary strings including any metacharacters we decide to use.
+			// The paired [] at least have some chance of being unambiguous.
+			extras = append(extras, fmt.Sprintf("[%s: %v]", attempts[i].ref.String(), attempts[i].err))
+		}
+		return nil, errors.Wrapf(primary.err, "(First tried mirrors: %s): %s", strings.Join(extras, "\n"), primary.ref.String())
```

Comment:
* Isn't `primary.ref.String()` already part of `primary.err`?
* "First tried mirrors:" is hard to understand. Maybe "Other contacted endpoints:"?

Reply:
- Not always, or at least not in full, see e.g. the “error pinging” example in https://github.com/containers/image/pull/788#issuecomment-572133188 .
- 
  - We don’t use the “endpoint” abstraction name in user-facing text so far; “mirror” directly corresponds to the configuration.
  - The “first tried” wording is attempting to convey order; “other contacted” feels to me a bit as if the primary was contacted first, and the others later. I do agree that this is hard to understand. Maybe I should give up on trying to explain the order in here?
  - (This is also somewhat constrained by the `errors.Wrapf` mechanism only adding text to the _start_ of the message. Ideally, it should be possible to say: “mirror 1 not found in DNS, LOOK HERE: mirror 2 found but authentication failed, primary not found in DNS” with both the appropriate order and the appropriate focus on the likely-to-be-important part, _and_ returning the correct typed error for “authentication failed”. This is anyway nowhere near that.)

Reply:
Just to throw something out there, maybe “Mirrors also failed:”?

Reply:
Sounds good, thanks!

Reply:
To be explicit, I’ll go with `Mirrors also failed:`, then.